Quiet "Broke affinity for irq" messages during shutdown

Quiet the "Broke affinity for irq" messages during a shutdown or
suspend, as it is already done during a reboot. Do this by conditionally
printing the messages depending on the system state, as originally
suggested by Jan when discussing commit e509b8e09c00 ("x86: fixup IRQs
when CPUs go down during shutdown").

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
Suggested-by: Jan Beulich <jbeulich@suse.com>

diff --git a/xen/arch/x86/include/asm/irq.h b/xen/arch/x86/include/asm/irq.h
index 7c825e9d9c0a..456e3c7f2a40 100644
--- a/xen/arch/x86/include/asm/irq.h
+++ b/xen/arch/x86/include/asm/irq.h
@@ -157,7 +157,7 @@ int map_domain_emuirq_pirq(struct domain *d, int pirq, int irq);
 int unmap_domain_pirq_emuirq(struct domain *d, int pirq);
 
 /* Reset irq affinities to match the given CPU mask. */
-void fixup_irqs(const cpumask_t *mask, bool verbose);
+void fixup_irqs(const cpumask_t *mask);
 void fixup_eoi(void);
 
 int  init_irq_data(void);
diff --git a/xen/arch/x86/irq.c b/xen/arch/x86/irq.c
index 67cbf6b979dc..fc6c27e109f4 100644
--- a/xen/arch/x86/irq.c
+++ b/xen/arch/x86/irq.c
@@ -2518,7 +2518,7 @@ static int __init setup_dump_irqs(void)
 __initcall(setup_dump_irqs);
 
 /* Reset irq affinities to match the given CPU mask. */
-void fixup_irqs(const cpumask_t *mask, bool verbose)
+void fixup_irqs(const cpumask_t *mask)
 {
     unsigned int irq;
     static int warned;
@@ -2622,7 +2622,8 @@ void fixup_irqs(const cpumask_t *mask, bool verbose)
 
         spin_unlock(&desc->lock);
 
-        if ( !verbose )
+        if ( system_state == SYS_STATE_reboot ||
+             system_state == SYS_STATE_suspend )
             continue;
 
         if ( !set_affinity )
diff --git a/xen/arch/x86/shutdown.c b/xen/arch/x86/shutdown.c
index acef03314372..950d621c7d56 100644
--- a/xen/arch/x86/shutdown.c
+++ b/xen/arch/x86/shutdown.c
@@ -572,6 +572,7 @@ void machine_restart(unsigned int delay_millisecs)
 
     if ( system_state >= SYS_STATE_smp_boot )
     {
+        system_state = SYS_STATE_reboot;
         local_irq_enable();
 
         /* Ensure we are the boot CPU. */
diff --git a/xen/arch/x86/smp.c b/xen/arch/x86/smp.c
index eef0f9c6cbf4..d6344a6a18dc 100644
--- a/xen/arch/x86/smp.c
+++ b/xen/arch/x86/smp.c
@@ -359,7 +359,7 @@ void smp_send_stop(void)
         int timeout = 10;
 
         local_irq_disable();
-        fixup_irqs(cpumask_of(cpu), 0);
+        fixup_irqs(cpumask_of(cpu));
         local_irq_enable();
 
         smp_call_function(stop_this_cpu, NULL, 0);
diff --git a/xen/arch/x86/smpboot.c b/xen/arch/x86/smpboot.c
index 08c0f2d9df04..d8b33589a712 100644
--- a/xen/arch/x86/smpboot.c
+++ b/xen/arch/x86/smpboot.c
@@ -1280,7 +1280,7 @@ void __cpu_disable(void)
 
     /* It's now safe to remove this processor from the online map */
     cpumask_clear_cpu(cpu, &cpu_online_map);
-    fixup_irqs(&cpu_online_map, 1);
+    fixup_irqs(&cpu_online_map);
     fixup_eoi();
 }
 
diff --git a/xen/include/xen/kernel.h b/xen/include/xen/kernel.h
index 8cd142032d3b..4f47a4eb3fa0 100644
--- a/xen/include/xen/kernel.h
+++ b/xen/include/xen/kernel.h
@@ -95,7 +95,8 @@ extern enum system_state {
     SYS_STATE_smp_boot,
     SYS_STATE_active,
     SYS_STATE_suspend,
-    SYS_STATE_resume
+    SYS_STATE_resume,
+    SYS_STATE_reboot,
 } system_state;
 
 bool_t is_active_kernel_text(unsigned long addr);
